trie/zk/merkle_tree.go

```diff
@@ -0,0 +1,482 @@
+// with the node that proves the absence of the key.
+func (t *MerkleTree) Prove(key []byte, fromLevel uint, writeNode func(TreeNode) error) error {
+	// notice Prove in secure tree "pass through" the key instead of secure it
+	// this keep consistent behavior with geth's secure trie
+	k, err := zkt.NewHashFromCheckedBytes(key)
+	if err == nil {
+		err = t.ComputeAllNodeHash(nil)
+	}
+	err = t.ComputeAllNodeHash(nil)
+	if err == nil {
+		err = t.ProveByPath(t.newTreePath(k), int(fromLevel), func(n TreeNode) error {
+			if leaf, ok := n.(*LeafNode); ok {
+				leaf.KeyPreimage = zkt.NewByte32FromBytesPaddingZero(key)
+			}
+			return writeNode(n)
+		})
+	}
+	return err
+}
+
+// ProveByPath constructs a merkle proof for SMT, it respect the protocol used by the ethereum-trie
+// but save the node data with a compact form
+func (t *MerkleTree) ProveByPath(path TreePath, fromLevel int, writeNode func(TreeNode) error) error {
+	for i, node := 0, t.rootNode; i < t.maxLevels; i++ {
+		if i >= fromLevel {
+			// TODO: notice here we may have broken some implicit on the proofDb:
+			// the key is not kecca(value) and it even can not be derived from the value by any means without a actually decoding
+			if err := writeNode(node); err != nil {
+				return err
+			}
+		}
+		switch n := node.(type) {
+		case *ParentNode:
+			node = t.getChild(n, path.Get(i))
+			continue
+		case *LeafNode:
+		case *EmptyNode:
+		case *HashNode:
+		default:
+			return trie.ErrInvalidNodeFound
+		}
+		return nil
+	}
+	return nil
+}
+
+func (t *MerkleTree) checkNodeHashConflictDuringUpdate(node TreeNode) error {
+	if t.dirtyNode == nil {
+		return nil
+	}
+	return t.checkNodeHashConflict(node, true, nil)
+}
+
+func (t *MerkleTree) checkNodeHashConflict(
+	node TreeNode,
+	updating bool,
+	handleDirtyNode func(node TreeNode) error,
+) error {
+	if t.dirtyNode == nil {
+		return nil
+	}
+	return computeNodeHash(node, func(node TreeNode) error {
+		hashString := string(node.Hash()[:])
+		if oldNode, ok := t.dirtyNode[hashString]; ok {
+			if bytes.Equal(oldNode.CanonicalValue(), node.CanonicalValue()) {
+				return nil
+			}
+			return trie.ErrNodeKeyAlreadyExists
+		}
+		if updating {
+			// Compute the hash during tree modification.
+			// If you don't copy, even if the hash is changed many times and stored in the map,
+			// the pointer hasn't changed, so you'll always be left with the last state of the node.
+			t.dirtyNode[hashString] = copyNode(node)
+		} else {
+			t.dirtyNode[hashString] = node
+		}
+		if handleDirtyNode == nil {
+			return nil
+		}
+		return handleDirtyNode(node)
+	})
+}
+
+// getChild If the child node is a hash node, decode it and update the parent node.
+func (t *MerkleTree) getChild(node *ParentNode, path byte) TreeNode {
+	if t.findBlobByHash == nil {
+		return node.Child(path)
+	}
+	if hashNode, ok := node.Child(path).(*HashNode); ok {
+		blob, _ := t.findBlobByHash(hashNode.Hash()[:])
```

Comment:
Why do you ignore the err here?

Reply:
because not finding the blob is not always an error case: if we don't find the blob, it will remain a hash node and the function using getChild will take care of it.
If I were to take a log there, it would likely be a lot of logs.

Reply:
Could you please a comment of this?